contrib/kafka/src/main/java/com/google/cloud/dataflow/contrib/kafka/KafkaSource.java

```diff
@@ -0,0 +1,806 @@
+   * Builds Unbounded Kafka Source.
+   *
+   * @param K key type
+   * @param V value type
+   */
+  public static class Builder<K, V> {
+
+    private List<String> topics;
+    // future: let users specify subset of partitions to read.
+    private Coder<K> keyCoder;
+    private Coder<V> valueCoder;
+    private SerializableFunction<KafkaRecord<K, V>, Instant> timestampFn = new NowTimestampFn<>();
+    private Optional<SerializableFunction<KafkaRecord<K, V>, Instant>> watermarkFn =
+        Optional.absent();
+
+    private Map<String, Object> mutableConsumerConfig = Maps.newHashMap();
+
+    /**
+     * set of properties that are not required or don't make sense for our consumer.
+     */
+    private static final Map<String, String> ignoredConsumerProperties = ImmutableMap.of(
+        ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, "Set keyDecoderFn instead",
+        ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, "Set valueDecoderFn instead"
+
+        // "group.id", "enable.auto.commit", "auto.commit.interval.ms" :
+        //     lets allow these, applications can have better resume point for restarts.
+        );
+
+    private Builder() {
+      // set config defaults
+      mutableConsumerConfig.putAll(ImmutableMap.of(
+          ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName(),
+          ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName(),
+          // default to latest offset when we are not resuming.
+          ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "latest",
+          // disable auto commit of offsets. we don't require group_id. could be enabled by user.
+          ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, false));
+    }
+
+    /**
+     * Set Kafka bootstrap servers (alternately, set "bootstrap.servers" Consumer property).
+     *
+     * @param bootstrapServers Bootstrap servers for Kafka.
+     * @return Builder
+     */
+    public Builder<K, V> withBootstrapServers(String bootstrapServers) {
+      return withConsumerProperty(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
+    }
+
+    /**
+     * Set Kafka topics to be consumed. This is required.
+     *
+     * @param topics topics to read from
+     * @return Builder
+     */
+    public Builder<K, V> withTopics(Collection<String> topics) {
+      this.topics = ImmutableList.copyOf(topics);
+      return this;
+    }
+
+    /**
+     * Set a {@link ConsumerConfig} configuration property.
+     *
+     * @param configKey configuration property name
+     * @param configValue value for configuration property
+     * @return Builder
+     */
+    public Builder<K, V> withConsumerProperty(String configKey, Object configValue) {
+      Preconditions.checkArgument(!ignoredConsumerProperties.containsKey(configKey),
+          "No need to configure '%s'. %s", configKey, ignoredConsumerProperties.get(configKey));
+      mutableConsumerConfig.put(configKey, configValue);
+      return this;
```

Comment:
we strongly discourage mutating these builders. Instead, it probably makes sense to just take a fully-formed `Map<String, Object>`.